warehouse/forklift/legacy.py

```diff
@@ -260,85 +260,66 @@ def _is_valid_dist_file(filename, filetype):
+                    )
+                    return False
+
+                # Check that the compression type is valid
+                for zinfo in zfp.infolist():
+                    if zinfo.compress_type not in {
+                        zipfile.ZIP_STORED,
+                        zipfile.ZIP_DEFLATED,
+                    }:
+                        return False
 
-        # Check that the compression type is valid
-        with zipfile.ZipFile(filename) as zfp:
-            for zinfo in zfp.infolist():
-                if zinfo.compress_type not in {
-                    zipfile.ZIP_STORED,
-                    zipfile.ZIP_DEFLATED,
-                }:
+                # Check that the right files are present
+                for zipname in zfp.namelist():
+                    _, tail = os.path.split(zipname)
+                    if filename.endswith(".zip") and tail == "PKG-INFO":
+                        break
+                    if filename.endswith(".whl") and tail == "WHEEL":
+                        break
+                else:
                     return False
 
-    if filename.endswith(".tar.gz"):
-        # TODO: Ideally Ensure the compression ratio is not absurd
-        # (decompression bomb), like we do for wheel/zip above.
+        except zipfile.BadZipFile:
+            return False
 
+    elif filename.endswith(".tar.gz"):
         # Ensure that this is a valid tar file, and that it contains PKG-INFO.
+        # TODO: Ideally Ensure the compression ratio is not absurd
+        # (decompression bomb), like we do for wheel/zip above.
         try:
             with tarfile.open(filename, "r:gz") as tar:
                 # This decompresses the entire stream to validate it and the
                 # tar within.  Easy CPU DoS attack. :/
                 bad_tar = True
                 member = tar.next()
                 while member:
-                    parts = os.path.split(member.name)
-                    if len(parts) == 2 and parts[1] == "PKG-INFO":
+                    _, tail = os.path.split(member.name)
+                    if tail == "PKG-INFO":
                         bad_tar = False
                     member = tar.next()
                 if bad_tar:
                     return False
         except (tarfile.ReadError, EOFError):
             return False
-    elif filename.endswith(".zip"):
-        # Ensure that the .zip is a valid zip file, and that it has a
-        # PKG-INFO file.
-        try:
-            with zipfile.ZipFile(filename, "r") as zfp:
-                for zipname in zfp.namelist():
-                    parts = os.path.split(zipname)
-                    if len(parts) == 2 and parts[1] == "PKG-INFO":
-                        # We need the no branch below to work around a bug in
-                        # coverage.py where it's detecting a missed branch
-                        # where there isn't one.
-                        break
```

Comment:
This comment is no longer, has the bug in coverage.py been fixed and we can remove a "no branch" (not sure what its referring to, is the comment referring to something that was there before?)

Reply:
I have no idea what this comment is talking about, coverage seems fine with this section.